### PR TITLE
8325643: G1: Refactor G1FlushHumongousCandidateRemSets

### PR DESCRIPTION
**Pull request by albertnetymk**
Simple refactoring to ensure both paths are checked regarding remset status.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8325643](https://bugs.openjdk.org/browse/JDK-8325643): G1: Refactor G1FlushHumongousCandidateRemSets (**Enhancement** - P4)

### Reviewers
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to [72e47bd5](https://git.openjdk.org/jdk/pull/17808/files/72e47bd5cb54a87621ef1989f0e16830bb27b3b8)
 * [Ivan Walulya](https://openjdk.org/census#iwalulya) (@walulyai - **Reviewer**) ⚠️ Review applies to [72e47bd5](https://git.openjdk.org/jdk/pull/17808/files/72e47bd5cb54a87621ef1989f0e16830bb27b3b8)
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**) ⚠️ Review applies to [72e47bd5](https://git.openjdk.org/jdk/pull/17808/files/72e47bd5cb54a87621ef1989f0e16830bb27b3b8)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17808/head:pull/17808` \
`$ git checkout pull/17808`

Update a local copy of the PR: \
`$ git checkout pull/17808` \
`$ git pull https://git.openjdk.org/jdk.git pull/17808/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17808`

View PR using the GUI difftool: \
`$ git pr show -t 17808`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17808.diff">https://git.openjdk.org/jdk/pull/17808.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17808#issuecomment-1938891973)

### CODE REVIEW

**Review by walulyai — APPROVED** [review]

File: src/hotspot/share/gc/g1/g1RemSet.cpp [hunk 1]
```diff
@@ -1174,24 +1174,26 @@ class G1MergeHeapRootsTask : public WorkerTask {
...
+        // (and probably never) we do not enter this path if there are other kind of
+        // remembered sets for this region.
+        // We want to continue collecting remembered set entries for humongous regions
+        // that were not reclaimed.
+        r->rem_set()->clear(true /* only_cardset */, true /* keep_tracked */);
+      }
 
-      assert(r->rem_set()->is_empty() && r->rem_set()->is_complete(), "must be for eager reclaim candidates");
+      assert(r->rem_set()->is_empty(), "must be empty after flushing");
+      assert(r->rem_set()->is_complete(), "should still be after flushing");
```
Comment (kimbarrett) [inline src/hotspot/share/gc/g1/g1RemSet.cpp:1196]: I think these asserts belong inside the if-not-empty body. Both have already
been checked already in the empty-case, and there's nothing to change them.
Only the flushing in the if-not-empty body could change them.

**Review by kimbarrett — CHANGES_REQUESTED** [review]

File: src/hotspot/share/gc/g1/g1RemSet.cpp (hunk 1 above)
Comment (albertnetymk) [reply src/hotspot/share/gc/g1/g1RemSet.cpp:1196]: The intention is to have them as postconditions for all humongous candidates, before exiting this method. I can add a `// Postcondition` there to make it explicit. Or, do you still prefer having them inside the if?

Comment (kimbarrett) [reply src/hotspot/share/gc/g1/g1RemSet.cpp:1196]: OK.  Such a comment would be helpful, at least to me.

**Review by kimbarrett — APPROVED** [review]

**Comment by albertnetymk** [comment]
Thanks for review.

/integrate

**Pull request closed without merging**